src/docker.c

```
@@ -0,0 +1,1754 @@
+    }
+    return num_containers;
+  }
+  if (node->u.array.len == 0) {
+    yajl_tree_free(node);
+    return 0;
+  }
+
+  const char *id_path[] = { "Id", (const char *) 0 };
+  if (YAJL_IS_ARRAY(node)) {
+    list = (char **) calloc(node->u.array.len, sizeof(char *));
+    if (list == NULL) {
+      ERROR("docker: extract_container_ids_from_response: malloc failed!");
+      yajl_tree_free(node);
+      return DOCKER_ERROR;
+    }
+    num_containers = node->u.array.len;
+    for (int i = 0; i < num_containers; i++) {
+      yajl_val elem = node->u.array.values[i];
+      yajl_val id_node = yajl_tree_get(elem, id_path, yajl_t_string);
+      if (!YAJL_IS_OBJECT(id_node)) {
+	list[i] = sstrdup(YAJL_GET_STRING(id_node));
+      } else {
+	ERROR("docker: Container ID could not be extracted.\n");
+        DEBUG("docker: Bad response:\n %s", response_buffer);
+      }
+    }
+    *(container_list) = list;
+  }
+  yajl_tree_free(node);
+  return num_containers;
+}
+
+/* Purpose: Queries the Docker API for the list of running containers and
+ *           stores them in a string list.
+ *
+ * Params: (const char *) socket: Docker API UNIX socket to connect to.
+ *          (const char *) version: Version of the Docker API to query.
+ *          (char ***) container_list: Pointer to a list of string holding
+ *                                      the container IDs after the function
+ *                                      returns.
+ * Returns: int
+ *          0 - if no running containers are found or an error occured.
+ *          Number of running containers.
+ **/
+static int get_container_list(const char *socket, const char *version,
+    char ***container_list) {
+  char *response_buffer = (char *) calloc(RESPONSE_BUFFER_SIZE, sizeof(char));
+  char *url = (char *) calloc(28, sizeof(char));
+  int count = 0;
+  if (response_buffer == NULL || url == NULL) {
+    ERROR("docker: get_container_list: malloc failed!");
```

Should we return an error here?